### PR TITLE
REGR: unpickling in 1.3.0 DataFrame created in 1.2.x #42345

### LINKED ISSUE #42345: BUG: regression for 1.3.0: saving a dataframe with the "pickle" package using pandas 1.2.5 would not open with "pickle" again with pandas 1.3.0
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

I created a notebook here to illustrate the issue

https://colab.research.google.com/drive/1por9_BTeLYvZxNzLOBv_jIsFh44mnz4N?usp=sharing

Main error message:

```
/usr/local/lib/python3.7/dist-packages/pandas/_libs/internals.pyx in pandas._libs.internals.BlockManager.__cinit__()

TypeError: __cinit__() takes at least 2 positional arguments (0 given)
```

#### Problem description

Dataframes saved with pandas 1.2.5 and "pickle" are no longer loadable with pandas 1.3.0

Related to #42343

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : f00ed8f47020034e752baf0250483053340971b0
python           : 3.7.10.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.4.104+
Version          : #1 SMP Sat Jun 5 09:50:34 PDT 2021
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.3.0
numpy            : 1.19.5
pytz             : 2018.9
dateutil         : 2.8.1
pip              : 19.3.1
setuptools       : 57.0.0
Cython           : 0.29.23
pytest           : 3.6.4
hypothesis       : None
sphinx           : 1.8.5
blosc            : None
feather          : 0.4.1
xlsxwriter       : None
lxml.etree       : 4.2.6
html5lib         : 1.0.1
pymysql          : None
psycopg2         : 2.7.6.1 (dt dec pq3 ext lo64)
jinja2           : 2.11.3
IPython          : 5.5.0
pandas_datareader: 0.9.0
bs4              : 4.6.3
bottleneck       : 1.3.2
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.2.2
numexpr          : 2.7.3
odfpy            : None
openpyxl         : 2.5.9
pandas_gbq       : 0.13.3
pyarrow          : 3.0.0
pyxlsb           : None
s3fs             : None
scipy            : 1.4.1
sqlalchemy       : 1.4.18
tables           : 3.4.4
tabulate         : 0.8.9
xarray           : 0.18.2
xlrd             : 1.1.0
xlwt             : 1.3.0
numba            : 0.51.2

</details>

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #42345
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/io/test_pickle.py [hunk 1]
```diff
@@ -635,3 +635,13 @@ def test_pickle_big_dataframe_compression(protocol, compression):
...
     tm.assert_frame_equal(df, result)
+
+
+def test_pickle_frame_v124_unpickle_130():
+    # GH#42345 DataFrame created in 1.2.x, unpickle in 1.3.x
+    path = os.path.join(legacy_dirname, "1.2.4", "empty_frame_v1_2_4-GH#42345.pkl")
+    with open(path, "rb") as fd:
+        df = pickle.load(fd)
+
+    expected = pd.DataFrame()
```
Comment (jreback) [inline pandas/tests/io/test_pickle.py:646]: can i add another case with columns and index (respectively) as well (but no rows)

File: doc/source/whatsnew/v1.3.1.rst [hunk 2]
```diff
@@ -15,6 +15,7 @@ including other versions of pandas.
 Fixed regressions
 ~~~~~~~~~~~~~~~~~
 - Pandas could not be built on PyPy (:issue:`42355`)
+- :class:`DataFrame` constructed in with an older version of pandas could not be unpickled (:issue:`42345`)
```
Comment (mzeitlin11) [inline doc/source/whatsnew/v1.3.1.rst:18]:
```suggestion
- :class:`DataFrame` constructed with an older version of pandas could not be unpickled (:issue:`42345`)
```

Comment (jbrockmendel) [reply doc/source/whatsnew/v1.3.1.rst:18]: good catch, will update

File: pandas/tests/io/test_pickle.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/tests/io/test_pickle.py:644]: this test is highly focused, let's hold off on adding to this test and do a dedicated pass using the generate_legacy_storage_files pattern

File: pandas/tests/io/test_pickle.py [hunk 3]
```diff
@@ -633,3 +633,13 @@ def test_pickle_big_dataframe_compression(protocol, compression):
         partial(pd.read_pickle, compression=compression),
     )
     tm.assert_frame_equal(df, result)
+
+
+def test_pickle_frame_v124_unpickle_130():
+    # GH#42345 DataFrame created in 1.2.x, unpickle in 1.3.x
+    path = os.path.join(legacy_dirname, "1.2.4", "empty_frame_v1_2_4-GH#42345.pkl")
+    with open(path, "rb") as fd:
```
Comment (jreback) [inline pandas/tests/io/test_pickle.py:641]: can you also do thsi with `pd.read_pickle` (I think it works w/o your change prob)

File: pandas/tests/io/test_pickle.py (hunk 1 above)
Comment (jreback) [reply pandas/tests/io/test_pickle.py:644]: ahh ok, yeah i think that would be a great idea.

**Comment by jreback** [comment]
@jbrockmendel ok i think then let's merge this and then backport the full testing of the 1.2.x pickles in 1.3?

**Comment by jbrockmendel** [comment]
> yah

**Comment by jreback** [comment]
@meeseeksdev backport 1.3.x

**Pull request merged**